README.md
service.downloadmanager.SickPotatoHead
======================================

SickPotatoHead is a fork of SABnzbd-Suite geared toward us unfortunate enough not to have newsgroups access. Updated frequently to newest versions of SB, CP, and HP.


service.downloadmanager.SickPotatoHead/changelog.txt
1.7.1
- update to SickRage-d8c2613
- update to CouchPotatoServer-0a97ed5
- update to HeadPhones-f205205

1.7.0
- update to SickRage-v2017.04.26-1
- update to CouchPotatoServer-711310d

1.6.9
- removed i686 biniaries
- update to SickRage-v2017.02.2-1
- update to CouchPotatoServer-b538f9a
- update to HeadPhones-v0.5.18
- update to lxml-3.7.1
- update to pyOpenSSL-16.2.0
- update to openssl-1.0.2j
- update to unrar-5.40
- added cryptography-1.7.1
- added enum34-1.1.6
- added cffi-1.9.1
- added libffi.so.6.0.4

1.6.8
- update to SickRage-8bf53e8
- update to CouchPotatoServer-4a0999d
- update to HeadPhones-v0.5.16

1.6.7
- update to SickRage-a70d27a
- update to CouchPotatoServer-98cc86d

1.6.6
- update to HeadPhones-v0.5.13
- update to SickRage-b790df7
- update to CouchPotatoServer-c4fad95

1.6.5
- update to HeadPhones-v0.5.11
- update to CouchPotatoServer-f94a8a9

1.6.4
- update to HeadPhones-v0.5.10

1.6.3
- update to SickRage-v5.1.2
- update to CouchPotatoServer-89fd4e1

1.6.2
- update to SickRage-v5.1

1.6.1
- update to SickRage-v4.2.1.07
- update to CouchPotatoServer-de141aa

1.6.0
- update to SickRage-v4.2.1.04

1.5.9
- update to SickRage-v4.2.0.6

1.5.8
- added i386 binaries for lxml
- update to SickRage-v4.1.0.2

1.5.7
- update to SickRage-v4.0.76.2
- removed mako, its now included in sickrage
- added lxml for couchpotato

1.5.6
- update to SickRage-v4.0.73.1

1.5.5
- fix for ssl errors
- update to SickRage-v4.0.73

1.5.4
- fix ssl
- update to SickRage-v4.0.69.1

1.5.3
- update to SickRage-v4.0.69
- patched py-unrar2 to work with unrar 5.30
- update to unrar 5.30 beta4

1.5.2
- update to SickRage-v4.0.66
- added distutils

1.5.1
- update to HeadPhones-v0.5.9
- update to SickRage-v4.0.60
- update to CouchPotatoServer-build/3.0.1

1.5.0
- update to HeadPhones-v0.5.5
- update to SickRage-v4.0.20
- update to CouchPotatoServer-c295b34

1.4.9
- update to SickRage-v4.0.12

1.4.8
- update to CouchPotatoServer-build/2.6.3
- update to HeadPhones-v0.5.4
- update to SickRage-2d8aad8

1.4.7
- updated translations
- update to HeadPhones-v0.5.2
- update to SickRage-v4.0.9

1.4.6
- update to sickrage-v3.9.2
- moved pidfiles to /var/run/

1.4.5
- removed bytecode for multiple interpreter compatibility
- added binaries for armv7l devices
- removed openelec requirement should work on all linux xbmc installs
- update to CouchPotatoServer-build/2.6.1
- update to HeadPhones-v0.5.1
- update to SickRage-v3.1.0

1.4.4
- update to CouchPotatoServer-2252ed7
- update to HeadPhones-6b4fbe0
- update to SickRage-9eb9400

1.4.3
- update to SickRage-ace2d3d

1.4.2
- update to CouchPotatoServer-04aa2e5
- update to HeadPhones-34097e2
- update to SickRage-6eccffc

1.4.1
- update to HeadPhones-5cb9d78
- fix for sickrage webui becoming inaccessable

1.4.0
- update to CouchPotatoServer-e75a852
- update to SickRage-dbc30c4

1.3.9
- switching to SickRage SickBeard branch

1.3.8
- Fix XBMC crash on frodo
- update to HeadPhones-c170b2a

1.3.7
- update to CouchPotatoServer-45b2dff

1.3.6
- update to CouchPotatoServer-b45307e
- update to HeadPhones-88af845

1.3.5
- Fix for installs not running transmission addon

1.3.4
- Missed some typos

1.3.3
- Fix for last update

1.3.2
- update to HeadPhones-e67d324
- Update tp CouchPotatoServer-2402834
- updates to addon structure

1.3.1
- update to CouchPotatoServer-b69f8b7
- Switching SickBeard to mr-orange branch-09ddc02
- added lxml
- several general updates

1.3.0
- update to CouchPotatoServer-d7445df

1.2.9
- update to CouchPotatoServer-2cfff73
- update to HeadPhones-e879af6

1.2.8
- update to CouchPotatoServer-190e1d2

1.2.7
- update to CouchPotatoServer-b468048
- bumped required python version to 2.1.0 for gotham
- bumped required openelec version to 3.0 (frodo)

1.2.6
- update to CouchPotatoServer-b427563
- update to SickBeard-9663671
- update to Headphones-8ffd451

1.2.5
- update to CouchPotatoServer-d764d0f

1.2.4
- update to SickBeard-c59063a
- update to Headphones-dc69421

1.2.3
- Attemping to fix unable to start error with 1.2.2

1.2.2
- Disable restarting suite opon exiting addon settings
- update to SickBeard-4d77db9

1.2.1
- update to SickBeard-55d9f11

service.downloadmanager.SickPotatoHead/service.py
#
from os import system
import resources.sickpotatohead as sickpotatohead
import xbmc

__scriptname__ = "SickPotatoHead"
__author__     = "lsellens"
__url__        = "https://github.com/lsellens/xbmc.addons"

# Launch programs
sickpotatohead.main()

class MyMonitor(xbmc.Monitor):
    def __init__(self, *args, **kwargs):
        xbmc.Monitor.__init__(self)

    def onSettingsChanged(self):
        system("kill `ps | grep -E 'python.*service.downloadmanager.SickPotatoHead' | awk '{print $1}'`")
        sickpotatohead.main()


if __name__ == '__main__':
    monitor = MyMonitor()
    
    while not monitor.abortRequested():
        if monitor.waitForAbort():
            # Shutdown SickPotatoHead
            system("kill `ps | grep -E 'python.*service.downloadmanager.SickPotatoHead' | awk '{print $1}'`")
            break



service.downloadmanager.SickPotatoHead/resources/SickPotatoHead.stop.py
import urllib2
import xbmc
import xbmcaddon
from lib.configobj import ConfigObj

__addon__ = xbmcaddon.Addon(id='service.downloadmanager.SickPotatoHead')
__addonhome__ = xbmc.translatePath(__addon__.getAddonInfo('profile'))
sickbeard_launch = (__addon__.getSetting('SICKBEARD_LAUNCH').lower() == 'true')
couchpotato_launch = (__addon__.getSetting('COUCHPOTATO_LAUNCH').lower() == 'true')
headphones_launch = (__addon__.getSetting('HEADPHONES_LAUNCH').lower() == 'true')

if sickbeard_launch:
    try:
        pSickBeardSettings = xbmc.translatePath(__addonhome__ + 'sickbeard.ini')
        sickBeardConfig = ConfigObj(pSickBeardSettings, create_empty=False)
        sickBeardApiKey = sickBeardConfig['General']['api_key']
        urllib2.urlopen('http://localhost:8082/api/' + sickBeardApiKey + '/?cmd=sb.shutdown')
        xbmc.log('AUDO: Shutting SickBeard down...', level=xbmc.LOGDEBUG)
    except Exception, e:
        xbmc.log('AUDO: SickBeard exception occurred', level=xbmc.LOGERROR)
        xbmc.log(str(e), level=xbmc.LOGERROR)
        pass

if couchpotato_launch:
    try:
        pCouchPotatoServerSettings = xbmc.translatePath(__addonhome__ + 'couchpotatoserver.ini')
        couchPotatoServerConfig = ConfigObj(pCouchPotatoServerSettings, create_empty=False, list_values=False)
        CouchPotatoApiKey = couchPotatoServerConfig['core']['api_key']
        urllib2.urlopen('http://localhost:8083/api/' + CouchPotatoApiKey + '/app.shutdown')
        xbmc.log('AUDO: Shutting CouchPotato down...', level=xbmc.LOGDEBUG)
    except Exception, e:
        xbmc.log('AUDO: CouchPotato exception occurred', level=xbmc.LOGERROR)
        xbmc.log(str(e), level=xbmc.LOGERROR)
        pass

if headphones_launch:
    try:
        pHeadphonesSettings = xbmc.translatePath(__addonhome__ + 'headphones.ini')
        headphonesConfig = ConfigObj(pHeadphonesSettings, create_empty=False)
        headphonesApiKey = headphonesConfig['General']['api_key']
        urllib2.urlopen('http://localhost:8084/api?apikey=' + headphonesApiKey + '&cmd=shutdown')
        xbmc.log('AUDO: Shutting HeadPhones down...', level=xbmc.LOGDEBUG)
    except Exception, e:
        xbmc.log('AUDO: HeadPhones exception occurred', level=xbmc.LOGERROR)
        xbmc.log(str(e), level=xbmc.LOGERROR)
        pass

service.downloadmanager.SickPotatoHead/resources/SickPotatoHead.py
# Initializes and launches Couchpotato V2, Sickbeard and Headphones
from xml.dom.minidom import parseString
from lib.configobj import ConfigObj
import os
import subprocess
import hashlib
import xbmc
import xbmcaddon
import xbmcvfs

# helper functions
# ----------------


def create_dir(dirname):
    if not xbmcvfs.exists(dirname):
        xbmcvfs.mkdirs(dirname)
        xbmc.log('SickPotatoHead: Created directory ' + dirname, level=xbmc.LOGDEBUG)
# define some things that we're gonna need, mainly paths
# ------------------------------------------------------

# addon
__addon__             = xbmcaddon.Addon(id='service.downloadmanager.SickPotatoHead')
__addonpath__         = xbmc.translatePath(__addon__.getAddonInfo('path'))
__addonhome__         = xbmc.translatePath(__addon__.getAddonInfo('profile'))

# settings
pDefaultSuiteSettings = xbmc.translatePath(__addonpath__ + '/settings-default.xml')
pSuiteSettings        = xbmc.translatePath(__addonhome__ + 'settings.xml')
pXbmcSettings         = xbmc.translatePath('special://home/userdata/guisettings.xml')
pSickBeardSettings    = xbmc.translatePath(__addonhome__ + 'sickbeard.ini')
pCouchPotatoServerSettings  = xbmc.translatePath(__addonhome__ + 'couchpotatoserver.ini')
pHeadphonesSettings   = xbmc.translatePath(__addonhome__ + 'headphones.ini')

# create the settings file if missing
if not xbmcvfs.exists(pSuiteSettings):
    xbmcvfs.copy(pDefaultSuiteSettings, pSuiteSettings)

#Get Device Home DIR
pHomeDIR = os.path.expanduser('~/')

# directories
pSickPotatoHeadComplete       = pHomeDIR+'downloads'
pSickPotatoHeadCompleteTV     = pHomeDIR+'downloads/tvshows'
pSickPotatoHeadCompleteMov    = pHomeDIR+'downloads/movies'
pSickPotatoHeadWatchDir       = pHomeDIR+'downloads/watch'

# pylib
pPylib                = xbmc.translatePath(__addonpath__ + '/resources/lib')

# service commands
sickBeard             = ['python', xbmc.translatePath(__addonpath__ + '/resources/SickBeard/SickBeard.py'),
                         '--daemon', '--datadir', __addonhome__,'--pidfile=/run/sickbeard.pid', '--config', pSickBeardSettings]
couchPotatoServer     = ['python', xbmc.translatePath(__addonpath__ + '/resources/CouchPotatoServer/CouchPotato.py'),
                         '--daemon', '--pid_file', xbmc.translatePath(__addonhome__ + 'couchpotato.pid'),
                         '--config_file', pCouchPotatoServerSettings]
headphones            = ['python', xbmc.translatePath(__addonpath__ + '/resources/Headphones/Headphones.py'),
                         '-d', '--datadir', __addonhome__, '--config', pHeadphonesSettings]

# create directories and settings if missing
# -----------------------------------------------

sbfirstLaunch = not xbmcvfs.exists(pSickBeardSettings)
cpfirstLaunch = not xbmcvfs.exists(pCouchPotatoServerSettings)
hpfirstLaunch = not xbmcvfs.exists(pHeadphonesSettings)

xbmc.log('SickPotatoHead: Creating directories if missing', level=xbmc.LOGDEBUG)
create_dir(__addonhome__)
create_dir(pSickPotatoHeadComplete)
create_dir(pSickPotatoHeadCompleteTV)
create_dir(pSickPotatoHeadCompleteMov)
create_dir(pSickPotatoHeadWatchDir)

# read addon and xbmc settings
# ----------------------------

# Transmission-Daemon
transauth = False
try:
    transmissionaddon = xbmcaddon.Addon(id='service.downloadmanager.transmission')
    transauth = (transmissionaddon.getSetting('TRANSMISSION_AUTH').lower() == 'true')

    if transauth:
        xbmc.log('SickPotatoHead: Transmission Authentication Enabled', level=xbmc.LOGDEBUG)
        transuser = (transmissionaddon.getSetting('TRANSMISSION_USER').decode('utf-8'))
        if transuser == '':
            transuser = None
        transpwd = (transmissionaddon.getSetting('TRANSMISSION_PWD').decode('utf-8'))
        if transpwd == '':
            transpwd = None
    else:
        xbmc.log('SickPotatoHead: Transmission Authentication Not Enabled', level=xbmc.LOGDEBUG)

except Exception, e:
    xbmc.log('SickPotatoHead: Transmission Settings are not present', level=xbmc.LOGNOTICE)
    xbmc.log(str(e), level=xbmc.LOGNOTICE)
    pass

# SickPotatoHead-Suite
user = (__addon__.getSetting('SICKPOTATOHEAD_USER').decode('utf-8'))
pwd = (__addon__.getSetting('SICKPOTATOHEAD_PWD').decode('utf-8'))
host = (__addon__.getSetting('SICKPOTATOHEAD_IP'))
sickbeard_launch = (__addon__.getSetting('SICKBEARD_LAUNCH').lower() == 'true')
couchpotato_launch = (__addon__.getSetting('COUCHPOTATO_LAUNCH').lower() == 'true')
headphones_launch = (__addon__.getSetting('HEADPHONES_LAUNCH').lower() == 'true')

# XBMC
fXbmcSettings                 = open(pXbmcSettings, 'r')
data                          = fXbmcSettings.read()
fXbmcSettings.close()
xbmcSettings                  = parseString(data)
xbmcServices                  = xbmcSettings.getElementsByTagName('services')[0]
xbmcPort                      = xbmcServices.getElementsByTagName('webserverport')[0].firstChild.data
try:
    xbmcUser                      = xbmcServices.getElementsByTagName('webserverusername')[0].firstChild.data
except StandardError:
    xbmcUser                      = ''
try:
    xbmcPwd                       = xbmcServices.getElementsByTagName('webserverpassword')[0].firstChild.data
except StandardError:
    xbmcPwd                       = ''

# prepare execution environment
# -----------------------------
parch                         = os.uname()[4]

xbmc.log('SickPotatoHead: ' + parch + ' architecture detected', level=xbmc.LOGDEBUG)

if not xbmcvfs.exists(xbmc.translatePath(pPylib + '/arch.' + parch)):
    xbmc.log('SickPotatoHead: Setting up binaries:', level=xbmc.LOGDEBUG)

    if xbmcvfs.exists(xbmc.translatePath(pPylib + '/arch.i686')):
        xbmcvfs.delete(xbmc.translatePath(pPylib + '/arch.i686'))
    if xbmcvfs.exists(xbmc.translatePath(pPylib + '/arch.x86_64')):
        xbmcvfs.delete(xbmc.translatePath(pPylib + '/arch.x86_64'))
    if xbmcvfs.exists(xbmc.translatePath(pPylib + '/arch.armv6l')):
        xbmcvfs.delete(xbmc.translatePath(pPylib + '/arch.armv6l'))
    if xbmcvfs.exists(xbmc.translatePath(pPylib + '/arch.armv7l')):
        xbmcvfs.delete(xbmc.translatePath(pPylib + '/arch.armv7l'))

    pnamemapper                   = xbmc.translatePath(pPylib + '/Cheetah/_namemapper.so')
    punrar                        = xbmc.translatePath(__addonpath__ + '/bin/unrar')

    try:
        fnamemapper                   = xbmc.translatePath(pPylib + '/multiarch/_namemapper.so.' + parch)
        xbmcvfs.copy(fnamemapper, pnamemapper)
        xbmc.log('SickPotatoHead: Copied _namemapper.so for ' + parch, level=xbmc.LOGDEBUG)
    except Exception, e:
        xbmc.log('SickPotatoHead: Error Copying _namemapper.so for ' + parch, level=xbmc.LOGERROR)
        xbmc.log(str(e), level=xbmc.LOGERROR)

    try:
        funrar                        = xbmc.translatePath(pPylib + '/multiarch/unrar.' + parch)
        xbmcvfs.copy(funrar, punrar)
        os.chmod(punrar, 0755)
        xbmc.log('SickPotatoHead: Copied unrar for ' + parch, level=xbmc.LOGDEBUG)
    except Exception, e:
        xbmc.log('SickPotatoHead: Error Copying unrar for ' + parch, level=xbmc.LOGERROR)
        xbmc.log(str(e), level=xbmc.LOGERROR)

    xbmcvfs.File(xbmc.translatePath(pPylib + '/arch.' + parch), 'w').close()

os.environ['PYTHONPATH'] = str(os.environ.get('PYTHONPATH')) + ':' + pPylib

# SickBeard start
try:
    # write SickBeard settings
    # ------------------------
    sickBeardConfig = ConfigObj(pSickBeardSettings, create_empty=True)
    defaultConfig = ConfigObj()
    defaultConfig['General'] = {}
    defaultConfig['General']['launch_browser']      = '0'
    defaultConfig['General']['version_notify']      = '0'
    defaultConfig['General']['web_port']            = '8082'
    defaultConfig['General']['web_host']            = host
    defaultConfig['General']['web_username']        = user
    defaultConfig['General']['web_password']        = pwd
    defaultConfig['General']['cache_dir']           = __addonhome__ + 'sbcache'
    defaultConfig['General']['log_dir']             = __addonhome__ + 'logs'
    defaultConfig['XBMC'] = {}
    defaultConfig['XBMC']['use_xbmc']               = '1'
    defaultConfig['XBMC']['xbmc_host']              = 'localhost:' + xbmcPort
    defaultConfig['XBMC']['xbmc_username']          = xbmcUser
    defaultConfig['XBMC']['xbmc_password']          = xbmcPwd
    defaultConfig['TORRENT'] = {}

    if transauth:
        defaultConfig['TORRENT']['torrent_username']         = transuser
        defaultConfig['TORRENT']['torrent_password']         = transpwd
        defaultConfig['TORRENT']['torrent_host']             = 'http://localhost:9091/'

    if sbfirstLaunch:
        defaultConfig['TORRENT']['torrent_path']          = pSickPotatoHeadCompleteTV
        defaultConfig['General']['tv_download_dir']       = pSickPotatoHeadComplete
        defaultConfig['General']['metadata_xbmc_12plus']  = '0|0|0|0|0|0|0|0|0|0'
        defaultConfig['General']['keep_processed_dir']    = '0'
        defaultConfig['General']['use_banner']            = '1'
        defaultConfig['General']['rename_episodes']       = '1'
        defaultConfig['General']['naming_ep_name']        = '0'
        defaultConfig['General']['naming_use_periods']    = '1'
        defaultConfig['General']['naming_sep_type']       = '1'
        defaultConfig['General']['naming_ep_type']        = '1'
        defaultConfig['General']['root_dirs']             = '0|'+pHomeDIR+'tvshows'
        defaultConfig['General']['naming_custom_abd']     = '0'
        defaultConfig['General']['naming_abd_pattern']    = '%SN - %A-D - %EN'
        defaultConfig['Blackhole'] = {}
        defaultConfig['Blackhole']['torrent_dir']         = pSickPotatoHeadWatchDir
        defaultConfig['EZRSS'] = {}
        defaultConfig['EZRSS']['ezrss']                   = '1'
        defaultConfig['PUBLICHD'] = {}
        defaultConfig['PUBLICHD']['publichd']             = '1'
        defaultConfig['KAT'] = {}
        defaultConfig['KAT']['kat']                       = '1'
        defaultConfig['THEPIRATEBAY'] = {}
        defaultConfig['THEPIRATEBAY']['thepiratebay']     = '1'
        defaultConfig['Womble'] = {}
        defaultConfig['Womble']['womble']                 = '0'
        defaultConfig['XBMC']['xbmc_notify_ondownload']   = '1'
        defaultConfig['XBMC']['xbmc_notify_onsnatch']     = '1'
        defaultConfig['XBMC']['xbmc_update_library']      = '1'
        defaultConfig['XBMC']['xbmc_update_full']         = '1'

    sickBeardConfig.merge(defaultConfig)
    sickBeardConfig.write()

    # launch SickBeard
    # ----------------
    if sickbeard_launch:
        xbmc.log('SickPotatoHead: Launching SickBeard...', level=xbmc.LOGDEBUG)
        subprocess.call(sickBeard, close_fds=True)
        xbmc.log('SickPotatoHead: ...done', level=xbmc.LOGDEBUG)
except Exception, e:
    xbmc.log('SickPotatoHead: SickBeard exception occurred', level=xbmc.LOGERROR)
    xbmc.log(str(e), level=xbmc.LOGERROR)
# SickBeard end

# CouchPotatoServer start
try:
    # empty password hack
    if pwd == '':
        md5pwd = ''
    else:
        #convert password to md5
        md5pwd = hashlib.md5(str(pwd)).hexdigest()

    # write CouchPotatoServer settings
    # --------------------------
    couchPotatoServerConfig = ConfigObj(pCouchPotatoServerSettings, create_empty=True, list_values=False)
    defaultConfig = ConfigObj()
    defaultConfig['core'] = {}
    defaultConfig['core']['username']               = user
    defaultConfig['core']['password']               = md5pwd
    defaultConfig['core']['port']                   = '8083'
    defaultConfig['core']['launch_browser']         = '0'
    defaultConfig['core']['host']                   = host
    defaultConfig['core']['data_dir']               = __addonhome__
    defaultConfig['updater'] = {}
    defaultConfig['updater']['enabled']             = '0'
    defaultConfig['updater']['notification']        = '0'
    defaultConfig['updater']['automatic']           = '0'
    defaultConfig['xbmc'] = {}
    defaultConfig['xbmc']['enabled']                = '1'
    defaultConfig['xbmc']['host']                   = 'localhost:' + xbmcPort
    defaultConfig['xbmc']['username']               = xbmcUser
    defaultConfig['xbmc']['password']               = xbmcPwd
    defaultConfig['transmission'] = {}

    if transauth:
        defaultConfig['transmission']['username']         = transuser
        defaultConfig['transmission']['password']         = transpwd
        defaultConfig['transmission']['host']             = 'localhost:9091'

    if cpfirstLaunch:
        defaultConfig['transmission']['directory']        = pSickPotatoHeadCompleteMov
        defaultConfig['xbmc']['xbmc_update_library']      = '1'
        defaultConfig['xbmc']['xbmc_update_full']         = '1'
        defaultConfig['xbmc']['xbmc_notify_onsnatch']     = '1'
        defaultConfig['xbmc']['xbmc_notify_ondownload']   = '1'
        defaultConfig['blackhole'] = {}
        defaultConfig['blackhole']['directory']           = pSickPotatoHeadWatchDir
        defaultConfig['blackhole']['use_for']             = 'torrent'
        defaultConfig['blackhole']['enabled']             = '0'
        defaultConfig['renamer'] = {}
        defaultConfig['renamer']['enabled']               = '0'
        defaultConfig['renamer']['from']                  = pSickPotatoHeadCompleteMov
        defaultConfig['renamer']['separator']             = '.'
        defaultConfig['renamer']['cleanup']               = '0'
        defaultConfig['nzbindex'] = {}
        defaultConfig['nzbindex']['enabled']              = '0'
        defaultConfig['mysterbin'] = {}
        defaultConfig['mysterbin']['enabled']             = '0'
        defaultConfig['core']['permission_folder']        = '0644'
        defaultConfig['core']['permission_file']          = '0644'
        defaultConfig['core']['show_wizard']            = '0'
        defaultConfig['core']['debug']                  = '0'
        defaultConfig['core']['development']            = '0'
        defaultConfig['searcher'] = {}
        defaultConfig['searcher']['preferred_method']     = 'torrent'

    couchPotatoServerConfig.merge(defaultConfig)
    couchPotatoServerConfig.write()

    # launch CouchPotatoServer
    # ------------------
    if couchpotato_launch:
        xbmc.log('SickPotatoHead: Launching CouchPotatoServer...', level=xbmc.LOGDEBUG)
        subprocess.call(couchPotatoServer, close_fds=True)
        xbmc.log('SickPotatoHead: ...done', level=xbmc.LOGDEBUG)
except Exception, e:
    xbmc.log('SickPotatoHead: CouchPotatoServer exception occurred', level=xbmc.LOGERROR)
    xbmc.log(str(e), level=xbmc.LOGERROR)
# CouchPotatoServer end

# Headphones start
try:
    # write Headphones settings
    # -------------------------
    headphonesConfig = ConfigObj(pHeadphonesSettings, create_empty=True)
    defaultConfig = ConfigObj()
    defaultConfig['General'] = {}
    defaultConfig['General']['launch_browser']            = '0'
    defaultConfig['General']['http_port']                 = '8084'
    defaultConfig['General']['http_host']                 = host
    defaultConfig['General']['http_username']             = user
    defaultConfig['General']['http_password']             = pwd
    defaultConfig['General']['check_github']              = '0'
    defaultConfig['General']['check_github_on_startup']   = '0'
    defaultConfig['General']['cache_dir']                 = __addonhome__ + 'hpcache'
    defaultConfig['General']['log_dir']                   = __addonhome__ + 'logs'
    defaultConfig['XBMC'] = {}
    defaultConfig['XBMC']['xbmc_enabled']                 = '1'
    defaultConfig['XBMC']['xbmc_host']                    = 'localhost:' + xbmcPort
    defaultConfig['XBMC']['xbmc_username']                = xbmcUser
    defaultConfig['XBMC']['xbmc_password']                = xbmcPwd
    defaultConfig['Transmission'] = {}

    if transauth:
        defaultConfig['Transmission']['transmission_username'] = transuser
        defaultConfig['Transmission']['transmission_password'] = transpwd
        defaultConfig['Transmission']['transmission_host']     = 'http://localhost:9091'

    if hpfirstLaunch:
        defaultConfig['XBMC']['xbmc_update']                  = '1'
        defaultConfig['XBMC']['xbmc_notify']                  = '1'
        defaultConfig['General']['music_dir']                 = pHomeDIR+'music'
        defaultConfig['General']['destination_dir']           = pHomeDIR+'music'
        defaultConfig['General']['torrentblackhole_dir']      = pSickPotatoHeadWatchDir
        defaultConfig['General']['download_torrent_dir']      = pSickPotatoHeadComplete
        defaultConfig['General']['move_files']                = '1'
        defaultConfig['General']['rename_files']              = '1'
        defaultConfig['General']['folder_permissions']        = '0644'
        defaultConfig['General']['isohunt']                   = '1'
        defaultConfig['General']['kat']                       = '1'
        defaultConfig['General']['mininova']                  = '1'
        defaultConfig['General']['piratebay']                 = '1'

    headphonesConfig.merge(defaultConfig)
    headphonesConfig.write()

    # launch Headphones
    # -----------------
    if headphones_launch:
        xbmc.log('SickPotatoHead: Launching Headphones...', level=xbmc.LOGDEBUG)
        subprocess.call(headphones, close_fds=True)
        xbmc.log('SickPotatoHead: ...done', level=xbmc.LOGDEBUG)
except Exception, e:
    xbmc.log('SickPotatoHead: Headphones exception occurred', level=xbmc.LOGERROR)
    xbmc.log(str(e), level=xbmc.LOGERROR)
# Headphones end
